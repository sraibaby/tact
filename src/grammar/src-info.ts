import { Interval as RawInterval } from "ohm-js";

export type ItemOrigin = "stdlib" | "user";

export interface LineAndColumnInfo {
    // offset: number;
    lineNum: number;
    colNum: number;
    // line: string;
    // prevLine: string;
    // nextLine: string;
    // toString(...ranges: number[][]): string;
}  

export type AbstractInterval = {
    contents: string;
    getLineAndColumnMessage(): string;
    getLineAndColumn(): LineAndColumnInfo;
    startIdx: number;
    endIdx: number;
};

export interface AbstractSrcInfo {
    get file(): string | null;
    get contents(): string;
    get interval(): AbstractInterval;
    get origin(): ItemOrigin;
}

/**
 * Information about source code location (file and interval within it)
 * and the source code contents.
 */
export class SrcInfo implements AbstractSrcInfo {
    readonly #interval: RawInterval;
    readonly #file: string | null;
    readonly #origin: ItemOrigin;

    constructor(
        interval: RawInterval,
        file: string | null,
        origin: ItemOrigin,
    ) {
        this.#interval = interval;
        this.#file = file;
        this.#origin = origin;
    }

    get file() {
        return this.#file;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }

    get origin() {
        return this.#origin;
    }
}