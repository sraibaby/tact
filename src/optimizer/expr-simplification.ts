import { CompilerContext } from "../context";
import { TactConstEvalError, throwInternalCompilerError } from "../errors";
import {
    AstConstantDef,
    AstContractDeclaration,
    AstExpression,
    AstFieldDecl,
    AstStatement,
    AstTraitDeclaration,
    AstValue,
    idText,
    SrcInfo,
} from "../grammar/ast";
import { Interpreter } from "../interpreter";
import { getType } from "../types/resolveDescriptors";
import { getExpType, registerExpType } from "../types/resolveExpression";
import { TypeRef, Value } from "../types/types";
import {
    OptimizationContext,
    registerAstNodeChange,
} from "./optimization-phase";
import { makeValueExpression, UnsupportedOperation } from "./util";

export function simplifyAllExpressions(optCtx: OptimizationContext) {
    // The interpreter in charge of simplifying expressions
    const interpreter = new Interpreter(optCtx.ctx);

    // Traverse the program and attempt to evaluate every expression

    for (const moduleItem of optCtx.modifiedAst.items) {
        switch (moduleItem.kind) {
            case "asm_function_def":
            case "native_function_decl":
            case "primitive_type_decl":
                // Nothing to simplify
                break;
            case "struct_decl":
            case "message_decl": {
                moduleItem.fields.forEach((field) => {
                    simplifyFieldDecl(field, optCtx, interpreter);
                });
                break;
            }
            case "constant_def": {
                simplifyConstantDef(moduleItem, optCtx, interpreter);
                break;
            }
            case "function_def": {
                moduleItem.statements.forEach((stmt) => {
                    simplifyStatement(stmt, optCtx, interpreter);
                });
                break;
            }
            case "contract": {
                moduleItem.declarations.forEach((decl) => {
                    simplifyContractDeclaration(decl, optCtx, interpreter);
                });
                break;
            }
            case "trait": {
                moduleItem.declarations.forEach((decl) => {
                    simplifyTraitDeclaration(decl, optCtx, interpreter);
                });
                break;
            }
            default:
                throwInternalCompilerError("Unrecognized module item kind");
        }
    }
}

function simplifyFieldDecl(
    ast: AstFieldDecl,
    optCtx: OptimizationContext,
    interpreter: Interpreter,
) {
    if (ast.initializer !== null) {
        ast.initializer = simplifyExpression(
            ast.initializer,
            optCtx,
            interpreter,
        );
    }
}

function simplifyConstantDef(
    ast: AstConstantDef,
    optCtx: OptimizationContext,
    interpreter: Interpreter,
) {
    ast.initializer = simplifyExpression(ast.initializer, optCtx, interpreter);
}

function simplifyContractDeclaration(
    decl: AstContractDeclaration,
    optCtx: OptimizationContext,
    interpreter: Interpreter,
) {
    switch (decl.kind) {
        case "asm_function_def": {
            // This kind is not changed by the optimizer
            break;
        }
        case "field_decl": {
            simplifyFieldDecl(decl, optCtx, interpreter);
            break;
        }
        case "constant_def": {
            simplifyConstantDef(decl, optCtx, interpreter);
            break;
        }
        case "function_def":
        case "receiver":
        case "contract_init": {
            decl.statements.forEach((stmt) => {
                simplifyStatement(stmt, optCtx, interpreter);
            });
            break;
        }
        default:
            throwInternalCompilerError(
                "Unrecognized contract declaration kind",
            );
    }
}

function simplifyTraitDeclaration(
    decl: AstTraitDeclaration,
    optCtx: OptimizationContext,
    interpreter: Interpreter,
) {
    switch (decl.kind) {
        case "asm_function_def":
        case "constant_decl":
        case "function_decl": {
            // These kinds are not changed by the optimizer
            break;
        }
        case "field_decl": {
            simplifyFieldDecl(decl, optCtx, interpreter);
            break;
        }
        case "constant_def": {
            simplifyConstantDef(decl, optCtx, interpreter);
            break;
        }
        case "function_def":
        case "receiver": {
            decl.statements.forEach((stmt) => {
                simplifyStatement(stmt, optCtx, interpreter);
            });
            break;
        }
        default:
            throwInternalCompilerError("Unrecognized trait declaration kind");
    }
}

function simplifyStatement(
    stmt: AstStatement,
    optCtx: OptimizationContext,
    interpreter: Interpreter,
) {
    switch (stmt.kind) {
        case "statement_assign":
        case "statement_expression":
        case "statement_let":
        case "statement_destruct":
        case "statement_augmentedassign": {
            stmt.expression = simplifyExpression(
                stmt.expression,
                optCtx,
                interpreter,
            );
            break;
        }
        case "statement_return": {
            if (stmt.expression !== null) {
                stmt.expression = simplifyExpression(
                    stmt.expression,
                    optCtx,
                    interpreter,
                );
            }
            break;
        }
        case "statement_condition": {
            stmt.condition = simplifyExpression(
                stmt.condition,
                optCtx,
                interpreter,
            );
            stmt.trueStatements.forEach((trueStmt) => {
                simplifyStatement(trueStmt, optCtx, interpreter);
            });

            if (stmt.falseStatements !== null) {
                stmt.falseStatements.forEach((falseStmt) => {
                    simplifyStatement(falseStmt, optCtx, interpreter);
                });
            }

            if (stmt.elseif !== null) {
                simplifyStatement(stmt.elseif, optCtx, interpreter);
            }
            break;
        }
        case "statement_foreach": {
            stmt.map = simplifyExpression(stmt.map, optCtx, interpreter);
            stmt.statements.forEach((loopStmt) => {
                simplifyStatement(loopStmt, optCtx, interpreter);
            });
            break;
        }
        case "statement_until":
        case "statement_while": {
            stmt.condition = simplifyExpression(
                stmt.condition,
                optCtx,
                interpreter,
            );
            stmt.statements.forEach((loopStmt) => {
                simplifyStatement(loopStmt, optCtx, interpreter);
            });
            break;
        }
        case "statement_repeat": {
            stmt.iterations = simplifyExpression(
                stmt.iterations,
                optCtx,
                interpreter,
            );
            stmt.statements.forEach((loopStmt) => {
                simplifyStatement(loopStmt, optCtx, interpreter);
            });
            break;
        }
        case "statement_try": {
            stmt.statements.forEach((tryStmt) => {
                simplifyStatement(tryStmt, optCtx, interpreter);
            });
            break;
        }
        case "statement_try_catch": {
            stmt.statements.forEach((tryStmt) => {
                simplifyStatement(tryStmt, optCtx, interpreter);
            });
            stmt.catchStatements.forEach((catchStmt) => {
                simplifyStatement(catchStmt, optCtx, interpreter);
            });
            break;
        }
        default:
            throwInternalCompilerError("Unrecognized statement kind");
    }
}

function simplifyExpression(
    expr: AstExpression,
    optCtx: OptimizationContext,
    interpreter: Interpreter,
): AstExpression {
    const value = tryExpressionSimplification(expr, interpreter);
    let newExpr = expr;
    if (typeof value !== "undefined") {
        try {
            newExpr = makeValueExpression(value, expr.loc);
            // Register the new expression in the context
            registerAstNodeChange(optCtx, expr, newExpr);
            // To maintain consistency with types in the CompilerContext, register the
            // types of all newly created expressions
            optCtx.ctx = registerAllSubExpTypes(
                optCtx.ctx,
                newExpr,
                getExpType(optCtx.ctx, expr),
            );
        } catch (e) {
            if (e instanceof UnsupportedOperation) {
                // This means that transforming the value into an AST node is
                // unsupported. Just use the original expression.
                newExpr = expr;
            } else {
                throw e;
            }
        }
    }

    return newExpr;
}

function tryExpressionSimplification(
    expr: AstExpression,
    interpreter: Interpreter,
): Value | undefined {
    try {
        // Eventually, this will be replaced by the partial evaluator.
        return interpreter.interpretExpression(expr);
    } catch (e) {
        if (e instanceof TactConstEvalError) {
            if (!e.fatal) {
                return undefined;
            }
        }
        throw e;
    }
}

function registerAllSubExpTypes(
    ctx: CompilerContext,
    expr: AstValue,
    expType: TypeRef,
): CompilerContext {
    switch (expr.kind) {
        case "boolean":
        case "number":
        case "string":
        case "null": {
            ctx = registerExpType(ctx, expr, expType);
            break;
        }
        case "struct_instance": {
            ctx = registerExpType(ctx, expr, expType);

            const structFields = getType(ctx, expr.type).fields;
            const fieldTypes: Map<string, TypeRef> = new Map();

            for (const field of structFields) {
                fieldTypes.set(field.name, field.type);
            }

            for (const fieldValue of expr.args) {
                const fieldType = fieldTypes.get(idText(fieldValue.field));
                if (typeof fieldType === "undefined") {
                    throwInternalCompilerError(
                        `Field ${idText(fieldValue.field)} does not have a declared type in struct ${idText(expr.type)}.`,
                        fieldValue.loc,
                    );
                }
                ctx = registerAllSubExpTypes(
                    ctx,
                    ensureAstValue(fieldValue.initializer, fieldValue.loc),
                    fieldType,
                );
            }
            break;
        }
        default:
            throwInternalCompilerError("Unrecognized AstValue.");
    }
    return ctx;
}

function ensureAstValue(expr: AstExpression, src: SrcInfo): AstValue {
    switch (expr.kind) {
        case "boolean":
        case "null":
        case "number":
        case "string":
        case "struct_instance":
            return expr;
        default:
            throwInternalCompilerError(
                `Expressions of kind ${expr.kind} are not ASTValues.`,
                src,
            );
    }
}
