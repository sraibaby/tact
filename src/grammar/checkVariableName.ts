import { throwCompilationError } from "../errors";
import { AbstractSrcInfo } from "./src-info";

export function checkVariableName(name: string, loc: AbstractSrcInfo) {
    if (name.startsWith("__gen")) {
        throwCompilationError(`Variable name cannot start with "__gen"`, loc);
    }
    if (name.startsWith("__tact")) {
        throwCompilationError(`Variable name cannot start with "__tact"`, loc);
    }
}
