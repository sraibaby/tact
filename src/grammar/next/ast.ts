/**
 * AST node constructors are not just constructors: they also generate ids
 * We have this file so that the "current id" state would not be stored globally
 */

import { Location } from "@langtools/pgen/runtime";
import * as A from "../ast";
import { AbstractSrcInfo } from "../src-info";

export const getAstSchema = (
    toSrcInfo: (location: Location) => AbstractSrcInfo,
    nextId: number = 1,
) => {
    const createAstNode = <T>(src: Omit<T, "id">): T => {
        return Object.freeze(Object.assign({ id: nextId++ }, src)) as T;
    };

    const cloneAstNode = <T extends { id: number }>(src: T): T => {
        return { ...src, id: nextId++ };
    };

    const constructors = {
        Module: (imports: A.AstImport[], items: A.AstModuleItem[]): A.AstModule => createAstNode<A.AstModule>({ kind: "module", imports, items, }),
        Import: (path: A.AstString, loc: Location): A.AstImport => createAstNode<A.AstImport>({ kind: "import", path, loc: toSrcInfo(loc), }),
        PrimitiveTypeDecl: (name: A.AstId, loc: Location): A.AstPrimitiveTypeDecl => createAstNode<A.AstPrimitiveTypeDecl>({ kind: "primitive_type_decl", name, loc: toSrcInfo(loc), }),
        FunctionDef: (attributes: A.AstFunctionAttribute[], name: A.AstId, retType: A.AstType | null, params: A.AstTypedParameter[], statements: A.AstStatement[], loc: Location): A.AstFunctionDef => createAstNode<A.AstFunctionDef>({ kind: "function_def", attributes, name, return: retType, params, statements, loc: toSrcInfo(loc), }),
        AsmFunctionDef: (shuffle: A.AstAsmShuffle, attributes: A.AstFunctionAttribute[], name: A.AstId, retType: A.AstType | null, params: A.AstTypedParameter[], instructions: readonly A.AstAsmInstruction[], loc: Location): A.AstAsmFunctionDef => createAstNode<A.AstAsmFunctionDef>({ kind: "asm_function_def", shuffle, attributes, name, return: retType, params, instructions, loc: toSrcInfo(loc), }),
        FunctionDecl: (attributes: A.AstFunctionAttribute[], name: A.AstId, retType: A.AstType | null, params: A.AstTypedParameter[], loc: Location): A.AstFunctionDecl => createAstNode<A.AstFunctionDecl>({ kind: "function_decl", attributes, name, return: retType, params, loc: toSrcInfo(loc), }),
        NativeFunctionDecl: (attributes: A.AstFunctionAttribute[], name: A.AstId, nativeName: A.AstFuncId, params: A.AstTypedParameter[], retType: A.AstType | null, loc: Location): A.AstNativeFunctionDecl => createAstNode<A.AstNativeFunctionDecl>({ kind: "native_function_decl", attributes, name, nativeName, params, return: retType, loc: toSrcInfo(loc), }),
        ConstantDef: (attributes: A.AstConstantAttribute[], name: A.AstId, type: A.AstType, initializer: A.AstExpression, loc: Location): A.AstConstantDef => createAstNode<A.AstConstantDef>({ kind: "constant_def", attributes, name, type, initializer, loc: toSrcInfo(loc), }),
        ConstantDecl: (attributes: A.AstConstantAttribute[], name: A.AstId, type: A.AstType, loc: Location): A.AstConstantDecl => createAstNode<A.AstConstantDecl>({ kind: "constant_decl", attributes, name, type, loc: toSrcInfo(loc), }),
        StructDecl: (name: A.AstId, fields: A.AstFieldDecl[], loc: Location): A.AstStructDecl => createAstNode<A.AstStructDecl>({ kind: "struct_decl", name, fields, loc: toSrcInfo(loc), }),
        MessageDecl: (name: A.AstId, opcode: A.AstNumber | null, fields: A.AstFieldDecl[], loc: Location): A.AstMessageDecl => createAstNode<A.AstMessageDecl>({ kind: "message_decl", name, opcode, fields, loc: toSrcInfo(loc), }),
        Contract: (name: A.AstId, traits: A.AstId[], attributes: A.AstContractAttribute[], declarations: A.AstContractDeclaration[], loc: Location): A.AstContract => createAstNode<A.AstContract>({ kind: "contract", name, traits, attributes, declarations, loc: toSrcInfo(loc), }),
        Trait: (name: A.AstId, traits: A.AstId[], attributes: A.AstContractAttribute[], declarations: A.AstTraitDeclaration[], loc: Location): A.AstTrait => createAstNode<A.AstTrait>({ kind: "trait", name, traits, attributes, declarations, loc: toSrcInfo(loc), }),
        FieldDecl: (name: A.AstId, type: A.AstType, initializer: A.AstExpression | null, as: A.AstId | null, loc: Location): A.AstFieldDecl => createAstNode<A.AstFieldDecl>({ kind: "field_decl", name, type, initializer, as, loc: toSrcInfo(loc), }),
        Receiver: (selector: A.AstReceiverKind, statements: A.AstStatement[], loc: Location): A.AstReceiver => createAstNode<A.AstReceiver>({ kind: "receiver", selector, statements, loc: toSrcInfo(loc), }),
        ContractInit: (params: A.AstTypedParameter[], statements: A.AstStatement[], loc: Location): A.AstContractInit => createAstNode<A.AstContractInit>({ kind: "contract_init", params, statements, loc: toSrcInfo(loc), }),
        StatementLet: (name: A.AstId, type: A.AstType | null, expression: A.AstExpression, loc: Location): A.AstStatementLet => createAstNode<A.AstStatementLet>({ kind: "statement_let", name, type, expression, loc: toSrcInfo(loc), }),
        StatementReturn: (expression: A.AstExpression | null, loc: Location): A.AstStatementReturn => createAstNode<A.AstStatementReturn>({ kind: "statement_return", expression, loc: toSrcInfo(loc), }),
        StatementExpression: (expression: A.AstExpression, loc: Location): A.AstStatementExpression => createAstNode<A.AstStatementExpression>({ kind: "statement_expression", expression, loc: toSrcInfo(loc), }),
        StatementAssign: (path: A.AstExpression, expression: A.AstExpression, loc: Location): A.AstStatementAssign => createAstNode<A.AstStatementAssign>({ kind: "statement_assign", path, expression, loc: toSrcInfo(loc), }),
        StatementAugmentedAssign: (op: A.AstAugmentedAssignOperation, path: A.AstExpression, expression: A.AstExpression, loc: Location): A.AstStatementAugmentedAssign => createAstNode<A.AstStatementAugmentedAssign>({ kind: "statement_augmentedassign", op, path, expression, loc: toSrcInfo(loc), }),
        Condition: (condition: A.AstExpression, trueStatements: A.AstStatement[], falseStatements: A.AstStatement[] | null, elseif: A.AstCondition | null, loc: Location): A.AstCondition => createAstNode<A.AstCondition>({ kind: "statement_condition", condition, trueStatements, falseStatements, elseif, loc: toSrcInfo(loc), }),
        StatementWhile: (condition: A.AstExpression, statements: A.AstStatement[], loc: Location): A.AstStatementWhile => createAstNode<A.AstStatementWhile>({ kind: "statement_while", condition, statements, loc: toSrcInfo(loc), }),
        StatementUntil: (condition: A.AstExpression, statements: A.AstStatement[], loc: Location): A.AstStatementUntil => createAstNode<A.AstStatementUntil>({ kind: "statement_until", condition, statements, loc: toSrcInfo(loc), }),
        StatementRepeat: (iterations: A.AstExpression, statements: A.AstStatement[], loc: Location): A.AstStatementRepeat => createAstNode<A.AstStatementRepeat>({ kind: "statement_repeat", iterations, statements, loc: toSrcInfo(loc), }),
        StatementTry: (statements: A.AstStatement[], loc: Location): A.AstStatementTry => createAstNode<A.AstStatementTry>({ kind: "statement_try", statements, loc: toSrcInfo(loc), }),
        StatementTryCatch: (statements: A.AstStatement[], catchName: A.AstId, catchStatements: A.AstStatement[], loc: Location): A.AstStatementTryCatch => createAstNode<A.AstStatementTryCatch>({ kind: "statement_try_catch", statements, catchName, catchStatements, loc: toSrcInfo(loc), }),
        StatementForEach: (keyName: A.AstId, valueName: A.AstId, map: A.AstExpression, statements: A.AstStatement[], loc: Location): A.AstStatementForEach => createAstNode<A.AstStatementForEach>({ kind: "statement_foreach", keyName, valueName, map, statements, loc: toSrcInfo(loc), }),
        TypeId: (text: string, loc: Location): A.AstTypeId => createAstNode<A.AstTypeId>({ kind: "type_id", text, loc: toSrcInfo(loc), }),
        OptionalType: (typeArg: A.AstType, loc: Location): A.AstOptionalType => createAstNode<A.AstOptionalType>({ kind: "optional_type", typeArg, loc: toSrcInfo(loc), }),
        MapType: (keyType: A.AstTypeId, keyStorageType: A.AstId | null, valueType: A.AstTypeId, valueStorageType: A.AstId | null, loc: Location): A.AstMapType => createAstNode<A.AstMapType>({ kind: "map_type", keyType, keyStorageType, valueType, valueStorageType, loc: toSrcInfo(loc), }),
        BouncedMessageType: (messageType: A.AstTypeId, loc: Location): A.AstBouncedMessageType => createAstNode<A.AstBouncedMessageType>({ kind: "bounced_message_type", messageType, loc: toSrcInfo(loc), }),
        OpBinary: (op: A.AstBinaryOperation, left: A.AstExpression, right: A.AstExpression, loc: Location): A.AstOpBinary => createAstNode<A.AstOpBinary>({ kind: "op_binary", op, left, right, loc: toSrcInfo(loc), }),
        OpUnary: (op: A.AstUnaryOperation, operand: A.AstExpression, loc: Location): A.AstOpUnary => createAstNode<A.AstOpUnary>({ kind: "op_unary", op, operand, loc: toSrcInfo(loc), }),
        FieldAccess: (aggregate: A.AstExpression, field: A.AstId, loc: Location): A.AstFieldAccess => createAstNode<A.AstFieldAccess>({ kind: "field_access", aggregate, field, loc: toSrcInfo(loc), }),
        MethodCall: (self: A.AstExpression, method: A.AstId, args: A.AstExpression[], loc: Location): A.AstMethodCall => createAstNode<A.AstMethodCall>({ kind: "method_call", self, method, args, loc: toSrcInfo(loc), }),
        StaticCall: (funcId: A.AstId, args: A.AstExpression[], loc: Location): A.AstStaticCall => createAstNode<A.AstStaticCall>({ kind: "static_call", function: funcId, args, loc: toSrcInfo(loc), }),
        StructInstance: (type: A.AstId, args: A.AstStructFieldInitializer[], loc: Location): A.AstStructInstance => createAstNode<A.AstStructInstance>({ kind: "struct_instance", type, args, loc: toSrcInfo(loc), }),
        StructFieldInitializer: (field: A.AstId, initializer: A.AstExpression, loc: Location): A.AstStructFieldInitializer => createAstNode<A.AstStructFieldInitializer>({ kind: "struct_field_initializer", field, initializer, loc: toSrcInfo(loc), }),
        InitOf: (contract: A.AstId, args: A.AstExpression[], loc: Location): A.AstInitOf => createAstNode<A.AstInitOf>({ kind: "init_of", contract, args, loc: toSrcInfo(loc), }),
        Conditional: (condition: A.AstExpression, thenBranch: A.AstExpression, elseBranch: A.AstExpression, loc: Location): A.AstConditional => createAstNode<A.AstConditional>({ kind: "conditional", condition, thenBranch, elseBranch, loc: toSrcInfo(loc), }),
        Id: (text: string, loc: Location): A.AstId => createAstNode<A.AstId>({ kind: "id", text, loc: toSrcInfo(loc), }),
        FuncId: (text: string, loc: Location): A.AstFuncId => createAstNode<A.AstFuncId>({ kind: "func_id", text, loc: toSrcInfo(loc), }),
        Null: (loc: Location): A.AstNull => createAstNode<A.AstNull>({ kind: "null", loc: toSrcInfo(loc), }),
        String: (value: string, loc: Location): A.AstString => createAstNode<A.AstString>({ kind: "string", value, loc: toSrcInfo(loc), }),
        Boolean: (value: boolean, loc: Location): A.AstBoolean => createAstNode<A.AstBoolean>({ kind: "boolean", value, loc: toSrcInfo(loc), }),
        Number: (base: A.AstNumberBase, value: bigint, loc: Location): A.AstNumber => createAstNode<A.AstNumber>({ kind: "number", base, value, loc: toSrcInfo(loc), }),
        ContractAttribute: (name: A.AstString, loc: Location): A.AstContractAttribute => createAstNode<A.AstContractAttribute>({ type: "interface", name, loc: toSrcInfo(loc), }),
        FunctionAttribute: (type: A.AstFunctionAttributeName, loc: Location): A.AstFunctionAttribute => ({ type, loc: toSrcInfo(loc), }),
        ConstantAttribute: (type: A.AstConstantAttributeName, loc: Location): A.AstConstantAttribute => ({ type, loc: toSrcInfo(loc), }),
        TypedParameter: (name: A.AstId, type: A.AstType, loc: Location): A.AstTypedParameter => createAstNode<A.AstTypedParameter>({ kind: "typed_parameter", name, type, loc: toSrcInfo(loc), }),
    };

    return {
        cloneAstNode,
        constructors,
    };
};

/**
 * List of all constructors for AST nodes
 */
export type AstSchema = ReturnType<typeof getAstSchema>["constructors"];