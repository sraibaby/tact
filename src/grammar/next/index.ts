import * as $ from '@langtools/pgen/runtime';
import * as G from './grammar';
import * as A from "../ast";
import { parseModule } from './semantics';
import { getAstSchema } from './ast';
import { getErrorSchema } from './errors';
import { AbstractSrcInfo, ItemOrigin } from '../src-info';

const throwParseError = (
    matchResult: MatchResult,
    path: string,
    origin: ItemOrigin,
): never => {
    const interval = matchResult.getInterval();
    const source = new SrcInfo(interval, path, origin);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const message = `Parse error: expected ${(matchResult as any).getExpectedText()}\n`;
    throw new TactParseError(
        `${locationStr(source)}${message}\n${interval.getLineAndColumnMessage()}`,
        source,
    );
}

const getSrcInfo = (path: string, origin: ItemOrigin) => {
    return (location: $.Location): AbstractSrcInfo => {
        
    };
};

export const getParser = () => {
    const parse = (
        src: string,
        path: string,
        origin: ItemOrigin,
    ): A.AstModule => {
        try {
            const toSrcInfo = getSrcInfo(path, origin);
            const astSchema = getAstSchema(toSrcInfo);
            const errorSchema = getErrorSchema();
            const context = {
                ast: astSchema.constructors,
                err: errorSchema.constructors,
            };
            const result = parseModule($.parse(G.Module, src))(context);
            const errors = errorSchema.reportedErrors;
            if (errors.length !== 0) {
                // TODO
            }
            return result;
        } catch (e) {
            if (e instanceof $.ParseError) {
                throwParseError(matchResult, path, origin);
            } else {
                throw e;
            }
        }
    }

    const parseExpression = (sourceCode: string): A.AstExpression => {
        const matchResult = tactGrammar.match(sourceCode, "Expression");
        if (matchResult.failed()) {
            throwParseError(matchResult, "", "user");
        }
        ctx = { origin: "user" };
        return semantics(matchResult).astOfExpression();
    };

    const parseImports = (
        src: string,
        path: string,
        origin: ItemOrigin,
    ): string[] => {
        return inFile(path, () => {
            return imports.children.map((item) => item.astOfImport());

            const matchResult = tactGrammar.match(src, "JustImports");
            if (matchResult.failed()) {
                throwParseError(matchResult, path, origin);
            }
            ctx = { origin };
            try {
                const imports: A.AstImport[] =
                    semantics(matchResult).astOfJustImports();
                return imports.map((imp) => imp.path.value);
            } finally {
                ctx = null;
            }
        });
    }

    return {
        parse,
        parseExpression,
        parseImports,
    };
};