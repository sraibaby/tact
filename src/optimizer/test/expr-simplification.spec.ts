import { featureEnable } from "../../config/features";
import { CompilerContext } from "../../context";
import { __DANGER_resetNodeId } from "../../grammar/ast";
import { openContext } from "../../grammar/store";
import { resolveDescriptors } from "../../types/resolveDescriptors";
import { getAllExpressionTypes } from "../../types/resolveExpression";
import { resolveStatements } from "../../types/resolveStatements";
import { loadCases } from "../../utils/loadCases";
import { simplifyAllExpressions } from "../expr-simplification";
import { prepareAstForOptimization } from "../optimization-phase";

describe("expression-simplification", () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    for (const r of loadCases(__dirname + "/success/")) {
        it("should pass expression simplification for " + r.name, () => {
            let ctx = openContext(
                new CompilerContext(),
                [{ code: r.code, path: "<unknown>", origin: "user" }],
                [],
            );
            ctx = featureEnable(ctx, "external");
            ctx = resolveDescriptors(ctx);
            ctx = resolveStatements(ctx);
            const optCtx = prepareAstForOptimization(ctx, true);
            simplifyAllExpressions(optCtx);
            expect(getAllExpressionTypes(optCtx.ctx)).toMatchSnapshot();
        });
    }
    for (const r of loadCases(__dirname + "/failed/")) {
        it("should fail expression simplification for " + r.name, () => {
            let ctx = openContext(
                new CompilerContext(),
                [{ code: r.code, path: "<unknown>", origin: "user" }],
                [],
            );
            ctx = featureEnable(ctx, "external");
            ctx = resolveDescriptors(ctx);
            ctx = resolveStatements(ctx);
            const optCtx = prepareAstForOptimization(ctx, true);
            expect(() => {
                simplifyAllExpressions(optCtx);
            }).toThrowErrorMatchingSnapshot();
        });
    }
});
