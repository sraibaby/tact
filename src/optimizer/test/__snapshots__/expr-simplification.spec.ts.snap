// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expression-simplification should fail expression simplification for interpreter-called-when-no-contract 1`] = `
"<unknown>:6:9: Cannot evaluate expression to a constant: divisor expression must be non-zero
Line 6, col 9:
  5 | fun blowup(v: Int) {
> 6 |     1 / v;
              ^
  7 | }
"
`;

exports[`expression-simplification should pass expression simplification for interpreter-simplifies-when-no-contract 1`] = `
[
  [
    "v",
    "Int",
  ],
  [
    "10",
    "Int",
  ],
  [
    "v + 10",
    "Int",
  ],
  [
    "3",
    "Int",
  ],
  [
    "v + 10 + 3",
    "Int",
  ],
  [
    "3",
    "Int",
  ],
  [
    "7",
    "Int",
  ],
  [
    "3 + 7",
    "Int",
  ],
  [
    "A {a: v + 10 + 3, b: 3 + 7}",
    "A",
  ],
  [
    "s",
    "A",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "B {nested: s, c: true}",
    "B",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exprFun1(0)",
    "A",
  ],
  [
    "exprFun1(0).a",
    "Int",
  ],
  [
    "2",
    "Int",
  ],
  [
    "exprFun1(2)",
    "A",
  ],
  [
    "exprFun2(exprFun1(2))",
    "B",
  ],
  [
    "1",
    "Int",
  ],
  [
    "exprFun1(1)",
    "A",
  ],
  [
    "exprFun2(exprFun1(1))",
    "B",
  ],
  [
    "exprFun2(exprFun1(1)).nested",
    "A",
  ],
  [
    "2",
    "Int",
  ],
  [
    "exprFun1(2)",
    "A",
  ],
  [
    "exprFun2(exprFun1(2))",
    "B",
  ],
  [
    "exprFun2(exprFun1(2)).c",
    "Bool",
  ],
  [
    "c1",
    "Int",
  ],
  [
    "exprFun1(c1)",
    "A",
  ],
  [
    "exprFun1(c1).a",
    "Int",
  ],
  [
    "0",
    "Int",
  ],
  [
    "exprFun1(c1).a > 0",
    "Bool",
  ],
  [
    "exprFun2(exprFun1(2)).c || exprFun1(c1).a > 0",
    "Bool",
  ],
  [
    "v + 10",
    "Int",
  ],
  [
    "v + 10 + 3",
    "Int",
  ],
  [
    "3 + 7",
    "Int",
  ],
  [
    "A {a: v + 10 + 3, b: 3 + 7}",
    "A",
  ],
  [
    "B {nested: s, c: true}",
    "B",
  ],
  [
    "exprFun1(0)",
    "A",
  ],
  [
    "exprFun1(0).a",
    "Int",
  ],
  [
    "exprFun1(2)",
    "A",
  ],
  [
    "exprFun2(exprFun1(2))",
    "B",
  ],
  [
    "exprFun1(1)",
    "A",
  ],
  [
    "exprFun2(exprFun1(1))",
    "B",
  ],
  [
    "exprFun2(exprFun1(1)).nested",
    "A",
  ],
  [
    "exprFun1(2)",
    "A",
  ],
  [
    "exprFun2(exprFun1(2))",
    "B",
  ],
  [
    "exprFun2(exprFun1(2)).c",
    "Bool",
  ],
  [
    "exprFun1(c1)",
    "A",
  ],
  [
    "exprFun1(c1).a",
    "Int",
  ],
  [
    "exprFun1(c1).a > 0",
    "Bool",
  ],
  [
    "exprFun2(exprFun1(2)).c || exprFun1(c1).a > 0",
    "Bool",
  ],
  [
    "13",
    "Int",
  ],
  [
    "B { nested: A { a: 15, b: 10 }, c: true }",
    "B",
  ],
  [
    "A { a: 15, b: 10 }",
    "A",
  ],
  [
    "15",
    "Int",
  ],
  [
    "10",
    "Int",
  ],
  [
    "true",
    "Bool",
  ],
  [
    "A { a: 14, b: 10 }",
    "A",
  ],
  [
    "14",
    "Int",
  ],
  [
    "10",
    "Int",
  ],
  [
    "true",
    "Bool",
  ],
]
`;
